addons/web_editor/static/lib/odoo-editor/src/OdooEditor.js

```diff
@@ -1484,10 +1485,17 @@ export class OdooEditor extends EventTarget {
         });
         if (!range) return;
         let { startContainer: start, startOffset, endContainer: end, endOffset } = range;
+        const startBlock = closestBlock(start);
+        const endBlock = closestBlock(end);
+        // Do not join blocks in the following cases:
+        // 1. start or end share a common ancestoral block with the range.
+        // 2. Selection spanning multiple TD's.
+        // 3. startBlock and endBlock are completely selected.
         const doJoin =
-            (closestBlock(start) !== closestBlock(range.commonAncestorContainer) ||
-            closestBlock(end) !== closestBlock(range.commonAncestorContainer))
-            && (closestBlock(start).tagName !== 'TD' && closestBlock(end).tagName !== 'TD');
+            (startBlock !== closestBlock(range.commonAncestorContainer) ||
+                endBlock !== closestBlock(range.commonAncestorContainer))
```

Comment:
While we're changing this, we might as well:

```suggestion
            !(startBlock === closestBlock(range.commonAncestorContainer) &&
                endBlock === closestBlock(range.commonAncestorContainer))
```